### ISSUE TITLE
test: browser.display.use_document_fonts

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#126]
see https://github.com/pyllyukko/user.js/pull/300

@nodiscc namely browser.display.use_document_fonts = 0 breaks glyphs/icon fonts. Examples: some uBo icons, icons in imgur.

Test 1 (vanilla Firefox)
- set fonts:`browser.display.use_document_fonts = 0`
- set glyphs: `gfx.downloadable_fonts.enabled = true`
- clear cache
- visit http://terrillthompson.com/tests/icomoon/

Windows results:  I can see the play, forward and stop buttons. I need to test this with glyphs set to false, to make sure that they are blocked. [EDIT: This icomoon is shown regardless of `gfx.downloadable_fonts.enabled` - tested in nilla FF. Need a better suite of tests]

I also need to find out what nodiscc means with imgur (I have never been to the site, still haven't as I type). I do know that with =0 and =true, I have zero issues on windows with font awesome, as per the link on the first line.

**Comment by Atavic** [comment]
On imgur I have no top left logo.

**Comment by RoxKilly** [comment]
`browser.display.use_document_fonts=0` no longer breaks icon fonts (including uBO icons) since Firefox 41, even though it still blocks the webpage from selecting standard fonts: [see ghacks](https://www.ghacks.net/2015/07/21/firefox-41-mozilla-modifies-how-icon-fonts-are-handled-by-the-browser/)

Prior to FF 41, this in fact broke uBo icons, but that has changed: [see this uBO issue](https://www.ghacks.net/2015/07/21/firefox-41-mozilla-modifies-how-icon-fonts-are-handled-by-the-browser/).  I've had this pref set to `0` for a couple of years and uBO icons display just fine here (Windows)

On the other hand, `gfx.downloadable_fonts.enabled=false` will break icon fonts, including uBO icons.

**Comment by Thorin-Oakenpants** [comment]
Yup, that's clear. I started this issue because nodiscc has raised irregularities (which is good, it can uncover stuff like that maxRichResults). In FF53 (debian) nodiscc is missing uBo icons despite having the correct pref values. I cannot replicate it. Part of this issue is we probably need to find out how FF decides something is a "glyph" and also how sites initiate the font (aren't there at least two ways to do this? css and @font-face?).

I will set up a test proc and results and screenshots (nilla FF) for others to replicate - just like we did for maxRich. We'll find the cause. Tests will include with and without uBo. I'll get to it the next day hopefully, unless someone else does it before me

**Comment by nodiscc** [comment]
> In FF53 (debian) nodiscc is missing uBo icons

FF 45.0.9 ESR

> how FF decides something is a "glyph"

There is nothing special about it, it's just a character in the `Common` unicode table, or even a simple ASCII character. For example on http://imgur.com/ the "cloud icon" (_New post_ button) is a simple `\5d` character (`view-source:http://s.imgur.com/min/global.css?1495492255`) which maps to a cloud icon in the character set of the `Open Sans` font they distribute (https://fonts.googleapis.com/css?family=Open+Sans:300,300i,400,400i,600,600i,700,700i,800,800i).

**Comment by Thorin-Oakenpants** [comment]
@nodiscc Yup. But Firefox has a special pref for glyphs, so my question is how does FF determine what it and isn't a glyph - *this* may explain some font icon behavior on various sites where what *we* see as an icon font, Firefox may not. But overall, "glyphs" as Firefox sees them are not blocked by document_fonts = 0.

And then we have this issue where your uBo is missing icons but I (and everyone else?) are not. I suspect this is a symptom of some linux or your profile quirk.

**Comment by earthlng** [comment]
I think the difference is installed vs downloadable (or bundled for extensions) fonts. That's why there are 2 prefs afaik

**Comment by Thorin-Oakenpants** [comment]
In order to solve what's happening - eg a site does not display glyphs, we first need to understand how Firefox decides if it is a glyph. - THE bugzilla ticket: https://bugzilla.mozilla.org/show_bug.cgi?id=789788

Note: Reading thru that ticket and links, leads me to some ideas about nodiscc's broken icons in uBo

### Problem:
- Icons are broken due to font suppression (i.e browser.display.use_document_fonts = 0)
  * I will refer to these "icon-fonts" as glyphs.
  * I will refer to "web-fonts" as downloadable fonts.
  * If a font is on your system, it can still be a glyph
  * It's impossible to determine if a font is a glyph with 100% certainty (at least in FF with rendering and downloading and all that goes on - technically maybe possible, but not feasible)
  * "the web has a large amount of cesspool dwelling hacks and bullshit developers" - Thorin
- Solution: gfx.downloadable_fonts.enabled = true
- Implementation: browser.display.use_document_fonts = 0 blocks all fonts except those font families specified, but allows an override if it is determined that the font is a "glyph"
- So HTF does FF work out what a glyph is? read on
- Interference: @font-face blocks, SVG/EOT/TTF/WOFF blocks, XSS blocks etc
- Note: It took them 3 and a 1/2 years to sort this out with a workable but not 100% solution

### Proposed Solutions

5 years ago

- glyphs *should* use the private-use area ([PUA](https://en.wikipedia.org/wiki/Private_Use_Areas))
  * "We use the Private Use (Unicode) range to avoid conflicts with actual letters." - https://github.com/styleguide/css/7.0
  * "Font Awesome uses the Unicode Private Use Area (PUA) to ensure screen readers do not read off random characters that represent icons" - https://github.com/FortAwesome/Font-Awesome/blob/master/less/font-awesome.less
- NOTE: lots (TONS) of glyphs do NOT use PUA codepoints for the icons - Thorin (cesspool and hacks)
- Allow any downloadable font specified by the page when it's used for INDIVIDUAL characters, and use the user-specified fonts otherwise
- A whitelist (and maybe blacklist)
   * This is both difficult to maintain and a bad experience for people writing content for the Web platform
- Accept all downloadable fonts would fix it (while degrading/altering the use-own-fonts option a bit)
   * more than a bit; it would make the preference no longer do what some users need it to do - not an option
- Add missing character fallback to the font load
   * I dont think it's desirable to individuate invalid characters (not feasible)
- Generalise/ establish a naming convention/pattern for glyphs, eg. accept fonts named ".*icon.*"
- NOTE: someone mentioned that in their studies so far they found about 90% of glyphs had *icon* in the font name - thorin
   * but we cant foretell what will happen in the future as glyphs become more common
- "allow_page_to_choose_TTF" and "allow_page_to_choose_SVG" (something to do with EOT, WOFF, TTF, and SVG formats used in some glyphs) - this would reduce the nummber of glyphs broken
   * eg github offers the octicon in all four formats, so it should be fixed by this

2 years later..

- The more complex the solutions we conceive here the less chance anyone who can code it will be able or inclined to do so
- `Foreach(webfont){ If ( any of the characters on the page in each webfont are invalid in the 'local-preference-set-font')`
   * Anything along these lines is pretty much doomed to failure. It would require analysis of the entire page... and/or might still be progressively downloading from the network, and/or might be constantly changing...

1 year later

1. give priority to the user's preferred fonts when the use_document_fonts option is off (0), but for characters not supported in the user's choice of font, use the page's specified fonts (whether local or downloaded or both) as usual, rather than going directly to system fallback fonts. This allows things like Github icons to continue working, as the PUA codepoints used should not be found in the user's pref fonts (normally), and so they'll fall back to the appropriate icon font from the page.
2. sites and fonts exist that misusing character codes as defined by Unicode, and the result is a lack of interoperability/compatibility
3. I don't see how the browser can "fix" this, except via some kind of explicit whitelist (or blacklist) of fonts/sites/etc, which would be a never-ending maintenance nightmare.
4. https://bugzilla.mozilla.org/show_bug.cgi?id=789788#c118 - explains it

So: no whitelists, no blacklists, no checking of font names for strings like `*icon*` etc - basically a fallback for fonts that use PUA

### Notes
- You can (always?) tell a PUA when its blocked by the 7-segment LCD-style debug codes in Bitstream Vera and DejaVu Sans (the rectangular icon that looks like a calulator!) - see imgur's missing glyphs are not these, the other test they are
- "The default Sans font on most Linux systems (DejaVu Sans) has glyphs for many of the PUA code points."
- NoScript can block @font-face which can be a vector of attack which can lead to font breakage including glyphs
- AdBlock (ewwww, I hope no-one is using it) can block/allow fonts
   * `*.ttf|`
   * `*.woff|`
   * `||fonts.googleapis.com^`

### Conclusion
- will do some tests soon and post back
- I started to look at imgur's glyphs in the top left nav - I just need to check the unicode code point, but I am 99% sure that these are not PUA
- The issue with uBo missing glyphs for nodiscc requires some tests in a nilla FF, but is not the same issue as imgur. Could be DejaVu Sans which has a ton of its own glyphs in PUA code points, which would explain the non calculator images shown - i.e nodiscc's default specified font does not fail those characters and displays the local specified characters - can be tested by changing the font set, clearing cache, and restarting (must restart to load font stuff). This is the exact behavior expected by the code.
- FF pref allows PUA glyphs only, not all glyphs - so maybe we should update the description

@nodiscc - what are you default fonts
 - see Options>Fonts & Colors>Advanced>Serif|Sans-serif|Monospace

PS: sorry for the long post and rambling, but if I wasted my time, you girls can spare 5 minutes, plus you might learn something (I did).

**Comment by Thorin-Oakenpants** [comment]
tl;dr:
-  5 years ago they suggested checking character unicode endpoints to see if they were PUA, and if they were to allow that font (local or downloaded) for that character
- 2.5 years later, they implemented a variation of that solution. They changed the behavior and wording of browser.display.use_document_fonts (flipped both) so that any fallback went to the user specified fonts. If the user-specified font is not using the endpoints then it will go get the font (assuming no other blocks like uBo, uM, NS, XSS, @font-face etc).
- MOST default fonts do not use PUA but some do. And lots of glyphs do use PUA, but not ALL, so is not 100% perfect
- "If the user-specified font is not using the endpoints" - DejaVu Sans on Linux may be an issue?

**Comment by Thorin-Oakenpants** [comment]
I think we need a better test, but here goes

Test1: https://www.w3schools.com/icons/tryit.asp?filename=tryicons_fa-list

- FF nilla or new profile
- `browser.display.use_document_fonts` = 0
- `gfx.downloadable_fonts.enabled` = false
- clear cache, quit & start (I did because I read something somewhere abut fonts loading on start)
- visit test page
- expected behavior: the w3schools "home" and "orientation" icons are displayed as unicode hex symbols (but the menu icon does display?)
- `gfx.downloadable_fonts.enabled` = true
- reload test page
- expected behavior - we get glyphs (home and orientation)

![glyph1](https://cloud.githubusercontent.com/assets/16656956/26390603/64ac53e6-40b3-11e7-97e9-a8cc6c9b2138.png)

If you do not see the glyphs, what are your default specified fonts?

Info: http://fontawesome.io/cheatsheet/
fa-home =  `[&#xf015;]`
fa-menu = IDK
fa-rotate = IDK

Why does fa-menu show, but not the others? Need some help here to get the unicode and check if which ones are PUA. Note: my specified fonts are Times New Roman, Arial, Courier New (Windows defaults)

@nodiscc : what do you get? This is not definitive, but we could easily code up a sample which outputs the uBo fonts that you can't see

**Comment by Thorin-Oakenpants** [comment]
IDK ... Using the test above, I simply changed the "try it" code, and see for yourself - glyphs missing from w3schools, but the code sample loads them (via cloudflare) as per the right hand side

![wtf](https://cloud.githubusercontent.com/assets/16656956/26395559/80dda892-40c4-11e7-9c4b-f8699fa27ab2.png)

**Comment by earthlng** [comment]
nice try with the explanation but totally confusing IMO. I think @nodiscc said the same thing but much easier to understand.
the characters used are (from left to right): \e800, \f0c9, \e813
the one you can see under Unicode is, as can be seen in the source code on the left: f03a
to make the hex codes into html characters you prepend the `&#x` and append a semicolon
if the numbers in the missing icons are not easy enough to read you can see them if you click "Inspect element" and switch to the Style Editor and look at the inline style sheet:
```css
.fa-home:before {content: '\e800'; }
.fa-save:before {content: '\e804'; }
.fa-rotate:before {content: '\e813'; }
.fa-menu:before { content: '\f0c9'; }
```

**Comment by earthlng** [comment]
uBO has 1 bundled font `fontawesome-webfont.ttf`, but IDK what the supported font formats are in Linux, maybe @nodiscc's Linux doesn't support TTF fonts and that's why he doesn't see the uBO icons.

**Comment by earthlng** [comment]
OT: frecency (frequency + recency) - from Martin's latest FF article. Thx Martin

**Comment by Atavic** [comment]
A good point [here](https://github.com/gorhill/uBlock/issues/605#issuecomment-132859267).
